ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,766 @@
+		state.calc_next_timestamp(header.timestamp(), self.period).map_err(|e| format!("{}", e))?;
+		self.block_state_by_hash.write().insert(header.hash(), state);
+
+		// no proof.
+		Ok(Vec::new())
+	}
+
+	// Our task here is to set difficulty
+	// TODO:(niklasad1): Return `Result<(), Error>` here instead
+	fn populate_from_parent(&self, header: &mut Header, parent: &Header) {
+		// TODO(https://github.com/paritytech/parity-ethereum/issues/10410): this is a horrible hack,
+		// it is due to the fact that enact and miner both use OpenBlock::new() which will both call
+		// this function. more refactoring is definitely needed.
+		if header.extra_data().len() < VANITY_LENGTH + SIGNATURE_LENGTH {
+			trace!(target: "engine", "populate_from_parent in sealing");
+
+			// It's unclear how to prevent creating new blocks unless we are authorized, the best way (and geth does this too)
+			// it's just to ignore setting an correct difficulty here, we will check authorization in next step in generate_seal anyway.
+			if let Some(signer) = self.signer.read().as_ref() {
+				let state = match self.state(&parent) {
+					Err(e) =>  {
+						trace!(target: "engine", "populate_from_parent: Unable to find parent state: {}, ignored.", e);
+						return;
+					}
+					Ok(state) => state,
+				};
+
+				if state.is_authorized(&signer.address()) {
+					if state.is_inturn(header.number(), &signer.address()) {
+						header.set_difficulty(DIFF_INTURN);
+					} else {
+						header.set_difficulty(DIFF_NOTURN);
+					}
+				}
+			} else {
+				trace!(target: "engine", "populate_from_parent: no signer registered");
+			}
+		}
+	}
+
+	fn set_signer(&self, signer: Box<EngineSigner>) {
+		trace!(target: "engine", "set_signer: {}", signer.address());
+		*self.signer.write() = Some(signer);
+	}
+
+	fn register_client(&self, client: Weak<EngineClient>) {
+		*self.client.write() = Some(client.clone());
+	}
+
+	fn step(&self) {
+		if self.signer.read().is_some() {
+			if let Some(ref weak) = *self.client.read() {
+				if let Some(c) = weak.upgrade() {
+					c.update_sealing();
+				}
+			}
+		}
+	}
+
+	fn stop(&mut self) {
+		if let Some(mut s) = self.step_service.as_mut() {
+			Arc::get_mut(&mut s).map(|x| x.stop());
+		}
+	}
+
+	/// Clique timestamp is set to parent + period , or current time which ever is higher.
+	fn open_block_header_timestamp(&self, parent_timestamp: u64) -> u64 {
+		let now = time::SystemTime::now().duration_since(time::UNIX_EPOCH).unwrap_or_default();
+		cmp::max(now.as_secs() as u64, parent_timestamp.saturating_add(self.period))
+	}
+
+	fn is_timestamp_valid(&self, header_timestamp: u64, parent_timestamp: u64) -> bool {
+		header_timestamp >= parent_timestamp.saturating_add(self.period)
+	}
+
+	fn fork_choice(&self, new: &ExtendedHeader, current: &ExtendedHeader) -> super::ForkChoice {
+		super::total_difficulty_fork_choice(new, current)
+	}
+
+	fn executive_author(&self, header: &Header) -> Address {
+		// Should have been verified now.
```

Comment:
This seems like a very handwavy assumption

Reply:
Yes, that comment is bad but the assumption is that `executive author` is called after the block has been verified https://github.com/paritytech/parity-ethereum/blob/455cd93a33661453a6c3afc8d6eb6b9f0f1b5a7a/ethcore/src/client/client.rs#L403

However, but I think `ec_recover` could still fail but I'm not really sure how might be better to return result here instead?

Reply:
https://github.com/goerli/parity-goerli/pull/82/files